### PR TITLE
Terraform output for workload reference

### PR DESCRIPTION
**Pull request by mofaizal**
## Overview/Summary

Replace this with a brief description of what this Pull Request fixes, changes, etc.

## This PR fixes/adds/changes/removes

1.  This example code helps you to extract required Terraform output values, which allow you to export structured data about your resources
2. The Outputs necessary to share data from a root module to child module and vice versa.

### Breaking Changes

1. N/A

## Testing Evidence

Please provide any testing evidence to show that your Pull Request works/fixes as described and planned (include screenshots, if appropriate).

## As part of this Pull Request I have

- [X] Checked for duplicate [Pull Requests](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/pulls)
- [ ] Associated it with relevant [issues](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/issues), for tracking and closure.
- [X] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/tree/main)
- [X] Performed testing and provided evidence.
- [X] Updated relevant and associated documentation.
- [ ] Updated the ["RELEASE"](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/issues) page with details needed for when this code change is released.

### CODE REVIEW

File: docs/wiki/Examples.md
Comment (krowlandson) [inline docs/wiki/Examples.md:33]: I think this would be best included within the `### Intermediate (Level 200)` examples.

Please also include in the `_Sidebar.md` for navigation in the Wiki

File: docs/wiki/Examples.md
Comment (krowlandson) [inline docs/wiki/Examples.md:56]: I'm not sure "Deploy" is the right verb for this guidance. Could we rename to something like `Use module outputs`?

File: docs/wiki/[Examples]-Deploy-Output-Values.md
Comment (krowlandson) [inline docs/wiki/[Examples]-Deploy-Output-Values.md:31]: Although we use this output in the test environments to give us a simple way to extract all of the resource IDs in a useful summary, I'm not sure this it the right approach for an example like this as it limits the data being output, and puts it in a very specific format.

Behind each resource type `key`, the module actually outputs the native Terraform resource outputs in map format, where they `key` represents the resource type

Instead I would rather see an example showing how to extract all config data for a specific resource type.

For example:

```hcl
output "azurerm_virtual_network" {
  value = values(module.test_connectivity.azurerm_virtual_network.connectivity)
}
```

Will extract a list of configuration settings for each virtual network created by the module, which will look like:

```shell
azurerm_virtual_network = [
  {
    "address_space" = tolist([
      "10.101.0.0/22",
    ])
    "bgp_community" = ""
    "ddos_protection_plan" = tolist([])
    "dns_servers" = tolist([])
    "edge_zone" = ""
    "flow_timeout_in_minutes" = 0
    "guid" = "ce03d850-478e-4603-9c03-a4ea6b9f5a18"
    "id" = "/subscriptions/fa2fa118-a60d-4700-9ef1-fa02beeaaea5/resourceGroups/12345-connectivity-northeurope/providers/Microsoft.Network/virtualNetworks/12345-hub-northeurope"
    "location" = "northeurope"
    "name" = "12345-hub-northeurope"
    "resource_group_name" = "12345-connectivity-northeurope"
    "subnet" = toset([])
    "tags" = tomap({
      "deployedBy" = "terraform/azure/caf-enterprise-scale/test_framework"
    })
    "timeouts" = null /* object */
  },
  {
    "address_space" = tolist([
      "10.100.0.0/22",
    ])
    "bgp_community" = ""
    "ddos_protection_plan" = tolist([])
    "dns_servers" = tolist([])
    "edge_zone" = ""
    "flow_timeout_in_minutes" = 0
    "guid" = "e535b12d-f92d-4e51-ab1f-09cc63681d8b"
    "id" = "/subscriptions/fa2fa118-a60d-4700-9ef1-fa02beeaaea5/resourceGroups/12345-connectivity-westeurope/providers/Microsoft.Network/virtualNetworks/12345-hub-westeurope"
    "location" = "westeurope"
    "name" = "12345-hub-westeurope"
    "resource_group_name" = "12345-connectivity-westeurope"
    "subnet" = toset([
      {
        "address_prefix" = "10.100.0.0/24"
        "id" = "/subscriptions/fa2fa118-a60d-4700-9ef1-fa02beeaaea5/resourceGroups/12345-connectivity-westeurope/providers/Microsoft.Network/virtualNetworks/12345-hub-westeurope/subnets/AzureFirewallSubnet"
        "name" = "AzureFirewallSubnet"
        "security_group" = ""
      },
    ])
    "tags" = tomap({
      "deployedBy" = "terraform/azure/caf-enterprise-scale/test_framework"
    })
    "timeouts" = null /* object */
  },
]
```

If you want to make this more useful for a customer to easily grab a network by type, you would want to make sure this is handled as a map, using a unique attribute to generate the key.

As we know the current key (the resource ID) isn't too friendly, but that we only allow a single virtual network hub per location, you might offer an example like the following:

```hcl
output "azurerm_virtual_network" {
  value = {
    for k,v in module.test_connectivity.azurerm_virtual_network.connectivity:
      v.location => v
    }
}
```

This would allow a customer to then reference a virtual network by location and extract all of the regular resource outputs and greatly simplify the examples you give below.

For example, the `local` for `connectivity_hub_vnet_name` would become:

```hcl
connectivity_hub_vnet_name = data.terraform_remote_state.eslz.outputs.azurerm_virtual_network.enterprise_scale.azurerm_virtual_network.westeurope.name
```

File: docs/wiki/[Examples]-Deploy-Output-Values.md
Comment (krowlandson) [inline docs/wiki/[Examples]-Deploy-Output-Values.md:13]: We should provide a complete example for ALZ, not just the outputs. If a customer doesn't enable `connectivity` resources, the examples below will not work.

File: docs/wiki/[Examples]-Deploy-Output-Values.md
Comment (krowlandson) [inline docs/wiki/[Examples]-Deploy-Output-Values.md:16]: When referring to the module, we try to use either `caf-enterprise-scale` or `Azure landing zones Terraform module`. Please can you use one of these for consistency?

We have also moved away from referencing "Enterprise-Scale" or "ESLZ", and prefer to use "Azure landing zones" or "ALZ". If you could update these references too, that would be great.

**Review by krowlandson — CHANGES_REQUESTED** [review]
Thank you for the PR @mofaizal ... some much needed content and I'm sure a LOT of people will find this useful 💥 

I've added some feedback which will need addressing before we can merge.

Please also take a look at the failed tests in the code review workflow and address these, as we cannot merge without these passing.

One point which I couldn't really put inline was whether we should use all of the output from the module via state file, or extract "key values" which would be useful for recommended module composition patterns like [dependency inversion](https://developer.hashicorp.com/terraform/language/modules/develop/composition#dependency-inversion). This would reduce the quantity of outputs needed from the module and is something I'm also discussing with someone else around how we make it easy to "chain modules" in a similar way to your proposal here. One option I'm considering is enriching and making the internal `template_file_variables` object an output, as this would help customers who are using the templating function to see what values they have available to use in template files, but also provide a pass-through capability for anything the customer wants to add to this object through the [template_file_variables](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/wiki/%5BVariables%5D-template_file_variables) input variable. using this approach, we could output crucial values (such as subscription ID, resource group name and resource name) which could then be used in a data resource to lookup any other required properties.

Other than that, I think I've covered off the main points in the feedback but please feel free to setup a call to discuss offline if you have any questions.

cc @matt-FFFFFF for further comments

File: docs/wiki/Examples.md
Comment (mofaizal) [reply docs/wiki/Examples.md:33]: Thank @krowlandson for the feedback, going on leave will look at the feedback when i back from leave :-)

Comment (krowlandson) [reply docs/wiki/Examples.md:33]: Thanks @mofaizal 👍🏻

**Comment by matt-FFFFFF** [comment]
@mofaizal please can you respond to the comments?

**Comment by mofaizal** [comment]
@matt-FFFFFF @krowlandson Thanks for the suggestion, working on revised version. Will update the document and create pull request again.

**Pull request closed without merging**
